erc20/vyper/vyper-erc20-spec.ini

```diff
@@ -131,60 +144,122 @@ requires:
     andBool 0 <=Int ALLOW     andBool ALLOW     <Int (2 ^Int 256)
 
 [transferFrom-success]
-k: #execute => #halt
 statusCode: _ => EVMC_SUCCESS
 output: _ => #asByteStackInWidth(1, 32)
 log: _:List ( .List => ListItem(#abiEventLog(ACCT_ID, "Transfer", #indexed(#address(FROM_ID)), #indexed(#address(TO_ID)), #uint256(VALUE))) )
 
 [transferFrom-success-1]
-storage:
-    #hashedLocation({COMPILER}, {_BALANCES},   FROM_ID)           |-> (BAL_FROM => BAL_FROM -Int VALUE)
-    #hashedLocation({COMPILER}, {_BALANCES},   TO_ID)             |-> (BAL_TO   => BAL_TO   +Int VALUE)
-    #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) |-> (ALLOW    => ALLOW    -Int VALUE)
-    _:Map
+storage: M1:IMap => M2:IMap
+origStorage: M1
 +requires:
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO
+    andBool select(M1, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW
     andBool FROM_ID =/=Int TO_ID
     andBool VALUE <=Int BAL_FROM
     andBool BAL_TO +Int VALUE <Int (2 ^Int 256)
     andBool VALUE <=Int ALLOW
 
+[transferFrom-success-1-a]
++requires:
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) =/=Int #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) =/=Int #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO +Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW -Int VALUE
+    andBool M1 ==IMap M2 except
+      (SetItem(#hashedLocation({COMPILER}, {_BALANCES}, FROM_ID))
+       SetItem(#hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) 
+       SetItem(#hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) .Set)
+
+[transferFrom-success-1-b]
++requires:
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) ==Int #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) =/=Int #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO +Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW -Int VALUE
```

Comment:
how could this possible where we have the hash collision?

Reply:
oh, BAL_FROM is equal to ALLOW

Reply:
i think it is better to explicitly say that BAL_FROM ==Int ALLOW in the ensures, to highlight it.